### PR TITLE
Require routing param for controlled pagination

### PR DESCRIPTION
**Pull request by zamoore**
### :pushpin: Summary

If merged, this PR will require that consumers of our pagination components (
`Hds::Pagination::Compact` & `Hds::Pagination::Numbered`) provide at least 1 routing argument 
(`@model`, `@models`, or `@route`) when routing is controlled externally.

### :hammer_and_wrench: Detailed description

- Added new assertions in component constructors that ensure the above
- Added some discriminated union types to the pagination type definitions that make it more clear when certain arguments are required.

### Questions

- How to we communicate this `required-if` behavior to the user in the docs? Is what we have enough?
- The discriminated union types seem to work in TS not with Glint. (IE, if I try to instantiate a pagination component with `@onPageChange` but omit routing params, I get no TS errors, but if I try to create an object with the same type in the component file, I get the expected errors.

### :link: External links

Jira ticket: [HDS-3439](https://hashicorp.atlassian.net/browse/HDS-3439)

***

### 👀 Component checklist

- [x] Percy was checked for any visual regression
- [x] A changelog entry was added via [Changesets](https://github.com/changesets/changesets) if needed (see [templates here](https://hashicorp.atlassian.net/wiki/spaces/HDS/pages/3243114706/Changelog+authoring+best+practices#Templates))

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-3439]: https://hashicorp.atlassian.net/browse/HDS-3439?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: website/docs/components/pagination/partials/code/how-to-use.md [hunk 1]
```diff
@@ -136,6 +136,12 @@ get demoQueryFunctionNumbered() {
 
 When the routing parameters are provided, the "navigation controls" are rendered as links and if the user clicks on one of them the page URL is automatically updated. This means that the component’s state is persisted **outside** of the component and so its whole state **must** be "controlled" by the consumer’s code (otherwise there would be conflicting states).
 
+
+!!! Info
```
Comment (shleewhite) [inline website/docs/components/pagination/partials/code/how-to-use.md:140]: Nit: should this be a warning? seems a little critical for people to know

File: showcase/tests/integration/components/hds/pagination/compact-test.js [hunk 2]
```diff
@@ -153,12 +153,30 @@ module('Integration | Component | hds/pagination/compact', function (hooks) {
...
-    await render(hbs`<Hds::Pagination::Compact @queryFunction="foo" />`);
+    await render(
+      hbs`<Hds::Pagination::Compact @queryFunction="foo" @model="test" />`
+    );
+    assert.throws(function () {
+      throw new Error(errorMessage);
+    });
+  });
+
+  test('it should throw an assertion if @queryFunction is provided without a routing argument', async function (assert) {
```
Comment (shleewhite) [inline showcase/tests/integration/components/hds/pagination/compact-test.js:169]: Is there a test that it doesnt throw an assertion if it is uncontrolled and doesn't have a routing argument?

thinking about it more, I guess if the other tests work then thats implied...

File: packages/components/src/components/hds/pagination/numbered/index.ts [hunk 3]
```diff
@@ -45,13 +50,46 @@ interface HdsPaginationNumberedArgs {
...
+  | HdsPaginationNumberedArgsControlledWithRoute;
+
+interface HdsPaginationNumberedArgsUncontrolled
+  extends HdsPaginationNumberedArgs {
+  queryFunction?: undefined;
+}
+
 export interface HdsPaginationNumberedSignature {
-  Args: HdsPaginationNumberedArgs & HdsPaginationRoutingProps;
+  Args:
```
Comment (shleewhite) [inline packages/components/src/components/hds/pagination/numbered/index.ts:90]: Nice!

**Review by shleewhite — APPROVED** [review]
Looks good! Just a couple nit comments.

File: showcase/tests/integration/components/hds/pagination/compact-test.js [hunk 4]
```diff
@@ -153,12 +153,30 @@ module('Integration | Component | hds/pagination/compact', function (hooks) {
 
   test('it should throw an assertion if @queryFunction is not a function', async function (assert) {
     const errorMessage =
-      '@queryFunction for "Hds::Pagination::Numbered" must be a function';
+      '@queryFunction for "Hds::Pagination::Compact" must be a function';
```
Comment (alex-ju) [inline showcase/tests/integration/components/hds/pagination/compact-test.js:156]: Ha! Good spot!

**Review by alex-ju — COMMENTED** [review]
Looks good overall. Clearer type abstractions! Aside from the warning I would also update the Component API section stating when `route/model/models` is required.

Would this change also apply to `replace` or only relevant to `route/model/models`?

File: packages/components/src/components/hds/pagination/numbered/index.ts (hunk 3 above)
Comment (zamoore) [reply packages/components/src/components/hds/pagination/numbered/index.ts:90]: Thanks!

File: showcase/tests/integration/components/hds/pagination/compact-test.js (hunk 4 above)
Comment (zamoore) [reply showcase/tests/integration/components/hds/pagination/compact-test.js:156]: Thanks!

**Comment by zamoore** [comment]
@alex-ju 

> Would this change also apply to `replace` or only relevant to `route/model/models`?

I don't believe that providing the replace argument should be required in any scenario. It should always be optional.

**Pull request merged**